src/core/lib/transport/promise_endpoint.h

```diff
@@ -0,0 +1,315 @@
+/*
+ * Copyright 2022 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef GRPC_CORE_LIB_TRANSPORT_PROMISE_ENDPOINT_H
+#define GRPC_CORE_LIB_TRANSPORT_PROMISE_ENDPOINT_H
+
+#include <grpc/support/port_platform.h>
+
+#include <stddef.h>
+#include <stdint.h>
+
+#include <functional>
+#include <memory>
+#include <utility>
+
+#include "absl/base/thread_annotations.h"
+#include "absl/status/status.h"
+#include "absl/status/statusor.h"
+#include "absl/types/optional.h"
+
+#include <grpc/event_engine/event_engine.h>
+#include <grpc/event_engine/slice.h>
+#include <grpc/event_engine/slice_buffer.h>
+#include <grpc/slice_buffer.h>
+#include <grpc/support/log.h>
+
+#include "src/core/lib/gprpp/sync.h"
+#include "src/core/lib/promise/activity.h"
+#include "src/core/lib/promise/poll.h"
+#include "src/core/lib/slice/slice.h"
+#include "src/core/lib/slice/slice_buffer.h"
+
+namespace grpc {
+
+namespace internal {
+
+class PromiseEndpoint {
+ public:
+  PromiseEndpoint(
+      std::unique_ptr<grpc_event_engine::experimental::EventEngine::Endpoint>
+          endpoint,
+      grpc_core::SliceBuffer already_received);
+  ~PromiseEndpoint();
+
+  auto Write(grpc_core::SliceBuffer data) {
```

Comment:
grpc_core::SliceBuffer may evolve over time to be different from grpc_event_engine::experimental::SliceBuffer. I think the input arguments to Read(..) and Write(..) must be of type grpc_event_engine::experimental::SliceBuffer

Reply:
Craig and I have discussed this. Please refer to the comments above.